[ROS] Add Official Docker image for ROS2 (#1381)
docker-library/docs: ros/content.md

# What is [ROS](https://index.ros.org/doc/ros2)?

The Robot Operating System (ROS) is a set of software libraries and tools that help you build robot applications. From drivers to state-of-the-art algorithms, and with powerful developer tools, ROS has what you need for your next robotics project. And it's all open source.

> [wikipedia.org/wiki/Robot_Operating_System](https://en.wikipedia.org/wiki/Robot_Operating_System)

[%%LOGO%%](https://index.ros.org/doc/ros2)

# How to use this image

## Creating a `Dockerfile` to install ROS packages

To create your own ROS docker images and install custom packages, here's a simple example of installing the C++, Python client library demos and security CLI using the official released Debian packages via apt-get.

```dockerfile
FROM %%IMAGE%%:crystal

# install ros packages for installed release
RUN apt-get update && apt-get install -y \
      ros-${ROS_DISTRO}-demo-nodes-cpp \
      ros-${ROS_DISTRO}-demo-nodes-py \
      ros-${ROS_DISTRO}-sros2 && \
    rm -rf /var/lib/apt/lists/*

# run ros package launch file
CMD ["ros2", "launch", "demo_nodes_cpp", "talker_listener.launch.py"]
```

> Note: all ROS images include a default entrypoint that sources the ROS environment setup before exiting the configured command, in this case the demo packages launch file. You can then build and run the Docker image like so:

Reviewer: exiting -> executing

> # What is [ROS](https://www.ros.org/)?

Reviewer 2: As these docs now default to ROS2, we could use this link instead http://www.ros2.org/ (which redirects to https://index.ros.org/doc/ros2/)

Author: Is that going to be the new landing page?
I just figured the top most link should point to the top most domain.

Author: I'm going to keep this as is. We can revisit this when updating the docs adding the ros1-bridge tag.

RUN vcs import ./ < ../overlay.repos

# copy manifests for caching
WORKDIR /opt
RUN mkdir -p /tmp/opt && \
    find ./ -name "package.xml" | \
      xargs cp --parents -t /tmp/opt && \
    find ./ -name "COLCON_IGNORE" | \
      xargs cp --parents -t /tmp/opt || true

# multi-stage for building
FROM $FROM_IMAGE AS builder

# install overlay dependencies
ARG OVERLAY_WS
WORKDIR $OVERLAY_WS
COPY --from=cacher /tmp/$OVERLAY_WS/src ./src
RUN . /opt/ros/$ROS_DISTRO/setup.sh && \
    apt-get update && rosdep install -y \
      --from-paths \
        src/ros2/demos/demo_nodes_cpp \
        src/ros2/demos/demo_nodes_py \
      --ignore-src \
    && rm -rf /var/lib/apt/lists/*

# build overlay source
COPY --from=cacher $OVERLAY_WS/src ./src
ARG OVERLAY_MIXINS="release"
RUN . /opt/ros/$ROS_DISTRO/setup.sh && \
    colcon build \
      --packages-select \
        demo_nodes_cpp \
        demo_nodes_py \
      --mixin $OVERLAY_MIXINS

# source entrypoint setup
ENV OVERLAY_WS $OVERLAY_WS
RUN sed --in-place --expression \
      '$isource "$OVERLAY_WS/install/setup.bash"' \
      /ros_entrypoint.sh

# run launch file
CMD ["ros2", "launch", "demo_nodes_cpp", "talker_listener.launch.py"]
```

> The example above starts by using [`vcstools`](https://github.com/vcstools/vcstools) to clone source repos of interest into the cacher stage. One could similarly `COPY` code from the local build context into the source directory as well. Package manifest files are then cached in a temporary directory where the following builder stage may copy from to install necessary dependencies with [`rosdep`](https://github.com/ros-infrastructure/rosdep). This is done prior copying the rest of the source files to preserve the multi stage build cache, given unaltered manifests do not altered declared dependencies, saving time and bandwidth. The overlay is then built using [`colcon`](https://colcon.readthedocs.io/en/released/), the entrypoint updated to source the workspace, and the default command set to launch the demo.

Reviewer 2: Nit^n: Can we make this one sentence per line ?

Author: See abf5b03

> The example above starts by using [`vcstools`](https://github.com/vcstools/vcstools) to clone source repos of interest into the cacher stage. One could similarly `COPY` code from the local build context into the source directory as well. Package manifest files are then cached in a temporary directory where the following builder stage may copy from to install necessary dependencies with [`rosdep`](https://github.com/ros-infrastructure/rosdep). This is done prior copying the rest of the source files to preserve the multi stage build cache, given unaltered manifests do not altered declared dependencies, saving time and bandwidth. The overlay is then built using [`colcon`](https://colcon.readthedocs.io/en/released/), the entrypoint updated to source the workspace, and the default command set to launch the demo.

Reviewer 2: ```suggestion
The example above starts by using [`vcstool`](https://github.com/dirk-thomas/vcstool) to clone source repos of interest into the cacher stage. One could similarly `COPY` code from the local build context into the source directory as well. Package manifest files are then cached in a temporary directory where the following builder stage may copy from to install necessary dependencies with [`rosdep`](https://github.com/ros-infrastructure/rosdep). This is done prior copying the rest of the source files to preserve the multi stage build cache, given unaltered manifests do not altered declared dependencies, saving time and bandwidth. The overlay is then built using [`colcon`](https://colcon.readthedocs.io/en/released/), the entrypoint updated to source the workspace, and the default command set to launch the demo.
```

We're actually using vcstool

> # What is [ROS](https://index.ros.org/doc/ros2)?

Reviewer 2: I understand the conclusion of merging repos and documentation, but is there a way to specify in some relevant places when the 2 versions vary. I'm thinking expecially here for documentation entry point and core licensing.

Author: How about splitting the the to into separate bullet points: https://github.com/docker-library/docs/pull/1381/commits/6d09179854033de0cc43bbf426832b327c6575f5

Author: Keeping ros org landing page as main link 966f6ae

[INFO] [talker]: Publishing: 'Hello World: 1'
[INFO] [listener]: I heard: [Hello World: 1]
[INFO] [talker]: Publishing: 'Hello World: 2'
[INFO] [listener]: I heard: [Hello World: 2]
...
```

## Creating a `Dockerfile` to build ROS packages

To create your own ROS docker images and build custom packages, here's a simple example of installing a package's build dependencies, compiling it from source, and installing the resulting build artifacts into a final multi-stage image layer.

```dockerfile
FROM %%IMAGE%%:dashing-ros-base AS builder

# clone package source
ENV ROS_WS /opt/ros_ws
RUN mkdir -p $ROS_WS/src
WORKDIR $ROS_WS
RUN git -C src clone \
      -b $ROS_DISTRO \
      https://github.com/ros2/demos.git

# install package dependencies
RUN . /opt/ros/$ROS_DISTRO/setup.sh && \
    apt-get update && rosdep update && \
    rosdep install -y \
      --from-paths \
        src/demos/demo_nodes_cpp \
      --ignore-src && \
    rm -rf /var/lib/apt/lists/*

# build package workspace
RUN . /opt/ros/$ROS_DISTRO/setup.sh && \
    colcon build \
      --packages-select \
        demo_nodes_cpp \
      --mixin release

# create new multi-stage

> FROM %%IMAGE%%:dashing-ros-core AS runner

Reviewer 2: Could we end up with missing runtme dependencies by having the runner being based on a different image than the builder?
Maybe the runner should use rosdep to install runtime dependencies?

Reviewer 3: @mikaelarguedas' question is actually partly why I posted [Generating ‘dev’ and runtime artefacts from ROS packages](https://discourse.ros.org/t/generating-dev-and-runtime-artefacts-from-ros-packages/12448).

Author: Looks like using rosdep to install runtime dependencies has been ticketed here:
https://github.com/ros-infrastructure/rosdep/pull/727

Once rosdep gets that feature, we could update example to use it. In the meantime, we'll just make a note to make sure the user is aware of need for runtime depencicies.

Reviewer 3: Doesn't this basically mean that these `Dockerfile`s are currently broken? Without the runtime dependencies, what's in `install` most likely doesn't run in the `runner` container.

Reviewer 2: > Once rosdep gets that feature, we could update example to use it. In the meantime, we'll just make a note to make sure the user is aware of need for runtime depencicies.

IMO this doesnt justify not installing runtime dependencies in the runner image. So I would argue that rosdep install should be used in the runner image and once rosdep allows to restrict the type of dependency to install, we can update the example to install "only runtime" dependencies.

These example dockerfiles work but only out of luck because the talker listener happen to have all their dependencies satisfied by `ros_core`. This will most likely not be the case for most use cases making this example confusing for users.

---

This also brings another point that is not explained here that is "why use ros-base a the builder base when ros-core is enough and consistent with the runner?"
This needs some comment explaining that the ros-core image doesnt come with colcon and other build tools anymore so ros_base needs to be used instead.

Author: > This needs some comment explaining that the ros-core image doesnt come with colcon and other build tools anymore so ros_base needs to be used instead.

Yep, this'll need to be updated.

> These example dockerfiles work but only out of luck because the talker listener happen to have all their dependencies satisfied by ros_core. This will most likely not be the case for most use cases making this example confusing for users.

What you alternative example package would you recommend? I only chose the talker and listener example as it's the classic demo, and its simple/quick to build, launch, and verify.

Reviewer 2: I think it's good to keep the `demo_nodes_cpp` package as the example :+1:

A good way to illustrate the need for installing runtime dependencies in the runner would be to launch the `add_two_ints.launch.py` example instead. As that uses insterfaces from `example_interfaces` that are not in ros_core, the container will fail.

Reviewer 2: In other words here is the runner I'm suggesting:

```
# create new multi-stage
FROM %%IMAGE%%:dashing-ros-core-bionic AS runner

# copy package install
ENV ROS_WS /opt/ros_ws
COPY --from=builder  $ROS_WS/install $ROS_WS/install

# install package dependencies
RUN apt-get update && apt-get install -y python3-rosdep && \
    apt-mark manual sudo && \
    rosdep init && \
    rosdep update --rosdistro $ROS_DISTRO && \
    . $ROS_WS/install/setup.sh && \
    rosdep install -y \
      --from-paths \
        $ROS_WS/install/share \
      --ignore-src && \
    apt-get remove -y python3-rosdep && apt-get autoremove -y && \
    rm -rf /var/lib/apt/lists/*

# source entrypoint setup
RUN sed --in-place --expression \
      '$isource "$ROS_WS/install/setup.bash"' \
      /ros_entrypoint.sh
# run launch file
CMD ["ros2", "launch", "demo_nodes_cpp", "add_two_ints.launch.py"]
```

Author: I think having to install rosdep back into ros-core is a little counter to its purpose. If rosdep could pipe a list of package names to a files that satisfy runtime requirements, then that file could be copied and piped back into apt-get in the runner image.

Although, this is still blocked by rosdep's feature set, so until then I've updated the docs to make careful note of this in 66047eb

Reviewer 2: > this is still blocked by rosdep's feature set

the rosdep feature set allows to do it. It's just not as optimized as it could be and we can easily add a flag once there is one to restrict the dependency set while preserving the logic of "automatically installing the dependencies you need" rather than having to figure it out and install them by hand.

I still think that this example is incorrect without it, either you're lucky and your dependencies are already in the base image, at which point rosdep is not necessary in the builder image either. Or (what I would guess is the vast majority of the cases) you need to install extra dependencies and ROS provides a tool for that called rosdep that we should use.

So I'm still in favor of using rosdep in the runner image as of now, uninstalling rosdep after was just an extra cleanup that may or may not be justified as rosdep has a pretty small footprint.

> If rosdep could pipe a list of package names to a files that satisfy runtime requirements, then that file could be copied and piped back into apt-get in the runner image.

As you use a different base image for building and running, the list of missing dependencies to install will differ between the builder or the runner so it doesnt seem bulletproof.

> helping to reconcile runtime vs development dependency discrepancies you may have otherwise forgotten to declare in your `package.xml` manifest

If the package.xml is not used to install dependencies, this approach cannot help catch missing or mismatching dependency declarations.

Author: > As you use a different base image for building and running, the list of missing dependencies to install will differ between the builder or the runner so it doesnt seem bulletproof.

That's where I think there should be a dry run option for rosdep where it just returns all the resolved runtime dependencies, regardless if they are installed or not, which would be complete enough for the runner to apt install. We do this for the CI in the navigation stack, but with a few more POSIX tools.

> If the package.xml is not used to install dependencies, this approach cannot help catch missing or mismatching dependency declarations.

That out of band dependency management then just falls onto the developer doesn't it?
rosdep can't help the developer if the developer doesn't also help rosdep.

Reviewer 2: > That out of band dependency management then just falls onto the developer doesn't it?
rosdep can't help the developer if the developer doesn't also help rosdep.

That's why I'm arguing for using rosdep in the example :wink: 
Otherwise this statement is misleading because the example does *not* allow catching mismatching dependency declarations because it ignores the declared dependencies (by not using rosdep)

> That's where I think there should be a dry run option for rosdep where it just returns all the resolved runtime dependencies

There is currently a dry run option, that lists all the installation commands it would run, (but it only lists the missing packages).
Is there an issue on rosdep to do the same for all (even already installed) packages ? it doesnt look like it would be a difficult feature to add.

Author: I don't know of such a ticket, but yeah, it should only need a small change to support. Then we can pipe those apt package names to a placeholder file to share across multistage.

Reviewer 2: > Then we can pipe those apt package names

I think piping the list of commands is better than the apt package names. Otherwise how would you handle other installers (e.g. pip packages)? trying to apt install those will likely not work.

> I don't know of such a ticket

I encourage you to open one if that's a feature you need and end up working around in other places.

In the meantime, can we agree on updating the example to install dependencies in the runner to move this forward? once extra rosdep features are available we can see how to update/optimize this example

Reviewer 3: The `--simulate` option of `rosdep` could *almost* do what you are asking for I believe.

See also [How can generate a list of dependencies from the src folder of a workspace?](https://answers.ros.org/question/251732) on ROS Answers.

Reviewer 2: :+1: yes that's the one I had in mind in:

> There is currently a dry run option, that lists all the installation commands it would run, (but it only lists the missing packages).

Author: > I encourage you to open one if that's a feature you need

On a closer search, looks like it's been open for while:

https://github.com/ros-infrastructure/rosdep/issues/587
https://github.com/ros-infrastructure/rosdep/pull/727

I've changed the multistage example to demo build and dependency caching with 2316f47 . We can always append a runtime stage later when rosdep is better up to the task. I think it best to keep the advance example clean and development oriented, so we can leave it at the builder stage to install exec dependencies for now.

# What is [ROS](https://www.ros.org/)?

The Robot Operating System (ROS) is a set of software libraries and tools that help you build robot applications.  
From drivers to state-of-the-art algorithms, and with powerful developer tools, ROS has what you need for your next robotics project.  

> And it's all open source.

Reviewer 4: I think you probably want to take a look at the "rich diff" in GitHub for this file -- using one-sentence-per-line results in literal line breaks in the output on GitHub (and Docker Hub's rendering is less consistent, which is why we tend to use one-line-per-paragraph even though it makes the diffs kind of annoying).

Author: I've reverted the one sentence per line commit.

With the advancements and standardization of software containers, roboticists are primed to acquire a host of improved developer tooling for building and shipping software. To help alleviate the growing pains and technical challenges of adopting new practices, we have focused on providing an official resource for using ROS with these new technologies.

For a complete listing of supported architecture and base images for each ROS Distribution Release, please read the official REP on target platforms for either [ROS 1](https://www.ros.org/reps/rep-0003.html) or for [ROS 2](https://www.ros.org/reps/rep-2000.html).

## Deployment suggestions

The available tags include supported distros along with a hierarchy tags based off the most common meta-package dependencies, designed to have a small footprint and simple configuration:

> -	`ros-core`: barebone ROS 2 install

Reviewer 2: If the goal is to conflate ROS 1 and 2 in this doc, is there a need to call out the version here?

Author: Switching to be version generic only when needed. 2cd312f

> -	`ros-base`: basic tools and libraries (also tagged with distro name with LTS version as `latest`)

Reviewer 2: This assumes the latest ROS 2 LTS is tagged as latest. This is not currently the case

Author: This will change with the new LTS foxy https://github.com/osrf/docker_images/pull/395

> The rest of the common meta-packages such as `desktop` and `ros1-bridge` are hosted on automatic build repos under OSRF's Docker Hub profile [here](https://hub.docker.com/r/osrf/ros/). These meta-packages include graphical dependencies and hook a host of other large packages such as X11, X server, etc. So in the interest of keep the official images lean and secure, the desktop packages are just be hosted with OSRF's profile.

Reviewer 2: Same here, ros1-bridge being ROS 2 only maybe its not worth pointing it out here.

Reviewer 2: Linking to the reps defining ROS variants would be useful here

Author: 50a2c27

With the advancements and standardization of software containers, roboticists are primed to acquire a host of improved developer tooling for building and shipping software. To help alleviate the growing pains and technical challenges of adopting new practices, we have focused on providing an official resource for using ROS with these new technologies.

> For a complete listing of supported architecture and base images for each ROS Distribution Release, please read the official REP on target platforms for either [ROS1](https://www.ros.org/reps/rep-0003.html) or for [ROS2](https://www.ros.org/reps/rep-2000.html).

Reviewer 2: ```suggestion
For a complete listing of supported architectures and base images for each ROS Distribution Release, please read the official REP on target platforms for either [ROS1](https://www.ros.org/reps/rep-0003.html) or for [ROS2](https://www.ros.org/reps/rep-2000.html).
```

Alternatively, more permissive network setting can be use to share all host network interfaces with the container, such as [`host` network driver](https://docs.docker.com/network/host/), simplifying connectivity with external network participants. Be aware however that this removes the networking namespace separation between containers, and can affect the ability of DDS participants communicate between containers, as documented [here](https://community.rti.com/kb/how-use-rti-connext-dds-communicate-across-docker-containers-using-host-driver).

## Deployment example

### Docker Compose

In this example we'll demonstrate using [`docker-compose`](https://docs.docker.com/compose/) to spawn a pair of message publisher and subscriber nodes in separate containers connected through shared software defined network.

> Create the directory `~/ros_demos` and add the first `Dockerfile` example from above. In the same directory, also create file `docker-compose.yml` with the following that runs a C++ publisher with a Python subscriber:

```yaml
version: '3'

services:
  talker:

>     build: ./Dockerfile

Reviewer 2: ```suggestion
    build:
      context: .
      dockerfile: Dockerfile
```

Reviewer 2: This is to avoid errors like:
```
  File "/usr/lib/python3/dist-packages/docker/api/build.py", line 142, in build
    raise TypeError("You must specify a directory to build in path")
TypeError: You must specify a directory to build in path
```

(on docker 19.03.2)

Author: Updated examples 66047eb

    command: ros2 run demo_nodes_cpp talker

  listener:

>     build: ./Dockerfile

Reviewer 2: ```suggestion
    build:
      context: .
      dockerfile: Dockerfile
```

Author: Updated examples 66047eb

    command: ros2 run demo_nodes_py listener
```

> Use docker-compose inside the same directory to launch our ROS nodes. Given the containers created derive from the same docker compose project, they will coexist on shared project network:

```console
$ docker-compose up -d
```

> Notice that a new network named `ros_demos` has been created, as can be shown further with:

```console
$ docker network inspect ros_demos
```

> We can monitor the logged output of each container, such as the listener node like so:

```console
$ docker-compose logs listener
```

> Finally, we can stop and remove all the relevant containers using docker-compose from the same directory:

```console
$ docker-compose stop
$ docker-compose rm
```

> Note: the auto-generated network, `ros_demos`, will persist until you explicitly remove it using `docker-compose down`.

> ### ROS 1 Bridge

Reviewer 2: Thanks for adding this :+1:

> To ease ROS 2 migration, [`ros1_bridge`](https://index.ros.org/p/ros1_bridge/github-ros2-ros1_bridge) is a ROS 2 package that provides bidirectional communication between ROS 1 and ROS 2. As a minimal example, given the ROS 2 Dockerfile above, we'll create the ROS 1 equivalent bellow, and name the Dockerfile appropriately. 

Reviewer 2: typo: bellow -> below

### Networks

ROS allows for peer-to-peer networking of processes (potentially distributed across machines) that are loosely coupled using the ROS communication infrastructure. ROS implements several different styles of communication, including synchronous RPC-style communication over services, asynchronous streaming of typed data over topics, combinations of both prior via request/reply and status/feedback over actions, and run-time settings via configuration over parameters. To abide by the best practice of [one process per container](https://docs.docker.com/articles/dockerfile_best-practices/), Docker networks can be used to string together several running ROS processes. For further details see the Deployment example further below.

> Alternatively, more permissive network setting can be use to share all host network interfaces with the container, such as [`host` network driver](https://docs.docker.com/network/host/), simplifying connectivity with external network participants. Be aware however that this removes the networking namespace separation between containers, and can affect the ability of DDS participants communicate between containers, as documented [here](https://community.rti.com/kb/how-use-rti-connext-dds-communicate-across-docker-containers-using-host-driver).

Reviewer 2: ```suggestion
Alternatively, more permissive network setting can be used to share all host network interfaces with the container, such as [`host` network driver](https://docs.docker.com/network/host/), simplifying connectivity with external network participants. Be aware however that this removes the networking namespace separation between containers, and can affect the ability of DDS participants to communicate between containers, as documented [here](https://community.rti.com/kb/how-use-rti-connext-dds-communicate-across-docker-containers-using-host-driver).
```
